python/lsst/daf/persistence/butler.py

```diff
@@ -229,35 +304,49 @@ def getKeys(self, datasetType=None, level=None):
         @returns (dict) valid data id keys; values are corresponding types.
         """
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        return self.repository.getKeys(datasetType, level)
+
+        keys = None
+        for repoData in self.inputs:
+            if not tag or len(tag.intersection(repoData.tags)) > 0:
```

Comment:
I think it should be an error if the same dataset type has different dataId keys in differently-tagged repos, but perhaps it's best to leave this flexibility in for possible future use.